Support annotating ForwardRef type

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68100

Currently annotation doesn't support ForwardRef type.

[ForwardRef](https://docs.python.org/3/library/typing.html) is a class used for internal typing representation of string forward references. For example, List["SomeClass"] is implicitly transformed into List[ForwardRef("SomeClass")]. This class should not be instantiated by a user, but may be used by introspection tools.

For example, `ForwardRef(Tensor)` is just a Tensor type and it should be annotated just as a Tensor type.

This pr will fetch the string representation and convert it to the correct type accordingly. It only cover `Tensor` type for now, and should be able to extend to other types easily

Differential Revision: [D32301442](https://our.internmc.facebook.com/intern/diff/D32301442/)

Commit:
Support annotating ForwardRef type

Currently annotation doesn't support ForwardRef type.

[ForwardRef](https://docs.python.org/3/library/typing.html) is a class used for internal typing representation of string forward references. For example, List["SomeClass"] is implicitly transformed into List[ForwardRef("SomeClass")]. This class should not be instantiated by a user, but may be used by introspection tools.

For example, `ForwardRef(Tensor)` is just a Tensor type and it should be annotated just as a Tensor type.

This pr will fetch the string representation and convert it to the correct type accordingly. It only cover `Tensor` type for now, and should be able to extend to other types easily

Differential Revision: [D32301442](https://our.internmc.facebook.com/intern/diff/D32301442/)

[ghstack-poisoned]

tugsbayasgalan commented:
Seems like ForwardRef is introduced in later pythons. So you might want to use conditional imports to fix the failing jobs.

Commit:
Update on "Support annotating ForwardRef type"


Currently annotation doesn't support ForwardRef type.

[ForwardRef](https://docs.python.org/3/library/typing.html) is a class used for internal typing representation of string forward references. For example, List["SomeClass"] is implicitly transformed into List[ForwardRef("SomeClass")]. This class should not be instantiated by a user, but may be used by introspection tools.

For example, `ForwardRef(Tensor)` is just a Tensor type and it should be annotated just as a Tensor type.

This pr will fetch the string representation and convert it to the correct type accordingly. It only cover `Tensor` type for now, and should be able to extend to other types easily

Differential Revision: [D32301442](https://our.internmc.facebook.com/intern/diff/D32301442/)

[ghstack-poisoned]

Commit:
Update on "Support annotating ForwardRef type"


Currently annotation doesn't support ForwardRef type.

[ForwardRef](https://docs.python.org/3/library/typing.html) is a class used for internal typing representation of string forward references. For example, List["SomeClass"] is implicitly transformed into List[ForwardRef("SomeClass")]. This class should not be instantiated by a user, but may be used by introspection tools.

For example, `ForwardRef(Tensor)` is just a Tensor type and it should be annotated just as a Tensor type.

This pr will fetch the string representation and convert it to the correct type accordingly. It only cover `Tensor` type for now, and should be able to extend to other types easily

Differential Revision: [D32301442](https://our.internmc.facebook.com/intern/diff/D32301442/)

[ghstack-poisoned]
